### PR TITLE
rec: Pass the Lua context to follow up queries (follow CNAME, dns64)

### LINKED ISSUE #10632: rec: policyEventFilter ignored by followCNAMERecords
- Program:  Recursor
 - Issue type: Bug report

### Short description
lua *policyEventFilter* method is not checked when recursor hits RPZ from "followCNAMERecords".

### Environment
 - Operating system: Ubuntu 20.04
 - Software version: PowerDNS Recursor 4.5.5
 - Software source: PowerDNS repo 

### Steps to reproduce
```bash
# recursor.conf
lua-config-file=/etc/powerdns/config.lua
lua-dns-script=/etc/powerdns/dns.lua

# config.lua
rpzFile('/etc/powerdns/rpz/list.rpz', { policyName="rpz1" })

# list.rpz
$TTL 10m;
$ORIGIN rpz1.
@               SOA localhost. root.localhost. ( 1 12h 15m 3w 2h)
@               NS localhost.  
restrict.youtube.com   CNAME .

# dns.lua
function preresolve(dq)
    dq:addAnswer(pdns.CNAME, "restrict.youtube.com")
    dq.followupFunction="followCNAMERecords"
    dq.rcode = pdns.NOERROR
    return true
end

function policyEventFilter(event)
  pdnslog(' = policyEventFilter fired')
  return false
end
```

### Expected behaviour

lua *policyEventFilter* method gets fired and "= policyEventFilter fired" log shows up

```bash
[2/1] question for 'restrict.youtube.com|A' from 10.1.1.13:36635 (proxied by 127.0.0.1:56102)
= policyEventFilter fired
restrict.youtube.com|A: RPZ Hit; PolicyName=rpz1; Trigger=restrict.youtube.com; Hit=restrict.youtube.com; Type=QName; Kind=.
[2/1] answer to question 'restrict.youtube.com|A': 0 answers, 1 additional, took 0 packets, 0 netw ms, 0 tot ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=3, dnssec=Indeterminate
```

### Actual behaviour

lua *policyEventFilter* method not running, no log

```bash
[2/1] question for 'www.youtube.com|A' from 10.1.1.13:51272 (proxied by 127.0.0.1:56102)
restrict.youtube.com|A: RPZ Hit; PolicyName=rpz1; Trigger=restrict.youtube.com; Hit=restrict.youtube.com; Type=QName; Kind=.
[2/1] answer to question 'www.youtube.com|A': 1 answers, 1 additional, took 0 packets, 0 netw ms, 0 tot ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=3, dnssec=Indeterminate
```

### Other information
After removing or commenting lua *preresolve* method, request is processed correctly and policyEventFilter method is checked like in excepted behaviour

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Fixes #10632.

Could use a regression test.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
Now with a regression test.

File: pdns/syncres.cc [hunk 1]
```diff
@@ -4533,18 +4533,21 @@ void SyncRes::parseEDNSSubnetAddFor(const std::string& subnetlist)
 }
 
 // used by PowerDNSLua - note that this neglects to add the packet count & statistics back to pdns_recursor.cc
-int directResolve(const DNSName& qname, const QType qtype, const QClass qclass, vector<DNSRecord>& ret)
+int directResolve(const DNSName& qname, const QType qtype, const QClass qclass, vector<DNSRecord>& ret, shared_ptr<RecursorLua4> pdl)
```
Comment (omoerbeek) [inline pdns/syncres.cc:4536]: To me it always feels a bit wrong to pass "large" objects by value. You might want to change to a ref here.

**Review by omoerbeek — APPROVED** [review]
Looks good, One nit, but OK either way.

File: pdns/syncres.cc (hunk 1 above)
Comment (rgacogne) [reply pdns/syncres.cc:4536]: I agree in principle, but in that specific case taking a reference makes it a bit harder that simply passing `nullptr` when there is no Lua context, for example in `sendNODLookup()`. As the performance impact should be negligible in that case, I'd rather keep it as it is.

**Pull request merged**